### PR TITLE
Add missing exchange binding for sharded queues

### PR DESCRIPTION
**Pull request by SwapnilBGaikwad**
For sharded queues we create N queues representing N shards.
All the sharded queues routing mapping is added in the rabbitmq exchange and we created sideline queue with it's exchange.

Routing key added in the binding for sideline exchange is only with base routing key such as `app.SEND_NOTIFICATION`.
This works for non sharded queues but sharded queues has routing defined as `app.SEND_NOTIFICATION_<shard_id>`.

Hence messages are getting dropped in the sideline exchange for sharded queue.
This MR add required binding in sideline to avoid any message drop from sideline.

Also have done few refactoring changes. Please review changes.

### CODE REVIEW

File: src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java [hunk 1]
```diff
@@ -150,6 +150,11 @@ public void ensure(final String queueName,
         log.info("Created queue: {} bound to {}", queueName, exchange);
     }
 
+    public void addBinding(String queueName, String exchange, String routingKey) throws Exception {
+        channel.queueBind(queueName, exchange, routingKey);
+        log.info("Created binding for queue : {} bound to {}", queueName, exchange);
```
Comment (vidhy) [inline src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java:155]: log routing key as well.

**Comment by vidhy** [comment]
Can we consider routing using wildcards (with topic exchange) instead of creating multiple bindings?

File: src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java (hunk 1 above)
Comment (SwapnilBGaikwad) [reply src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java]: Done.

**Comment by SwapnilBGaikwad** [comment]
Reason for not using wild card instead of direct mapping is currently topic exchanges only support # and * which are work based.
Since sharing is done using separator `_` word wild card does not work.

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java [hunk 2]
```diff
@@ -117,16 +117,18 @@ public void start() throws Exception {
         ensureExchange(dlx);
 
         this.publishChannel = connection.newChannel();
-        connection.ensure(queueName + "_SIDELINE", queueName, dlx,
-                connection.rmqOpts(config));
+        String sidelineQueueName = NamingUtils.getSideline(queueName);
         if (config.isSharded()) {
             int bound = config.getShardCount();
+            connection.ensure(sidelineQueueName, queueName, dlx, connection.rmqOpts(config));
```
Comment (r0goyal) [inline src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java:123]: Same line is present in if else block. Extract it out

**Comment by r0goyal** [comment]
@SwapnilBGaikwad Add unit test cases for the changes. Rest looks good

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java (hunk 2 above)
Comment (SwapnilBGaikwad) [reply src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java]: Done.

**Review by santanusinha — APPROVED** [review]

File: src/test/java/io/appform/dropwizard/actors/connectivity/actor/RMQTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import javax.validation.Validation;
+import java.util.concurrent.Executors;
+
+public class RMQTest {
+    private static ConnectionRegistry connectionRegistry;
```
Comment (r0goyal) [inline src/test/java/io/appform/dropwizard/actors/connectivity/actor/RMQTest.java:35]: Refer `NamespacedQueuesTest` for mode of writing tests

File: src/test/java/io/appform/dropwizard/actors/connectivity/actor/NamespacedQueuesTest.java [hunk 4]
```diff
@@ -177,6 +244,24 @@ private Response sendRequest(String endpoint) {
...
+                .header("Authorization", credential)
+                .post(RequestBody.create(body.getBytes(StandardCharsets.UTF_8)))
+                .build();
+
+        try {
+            return client.newCall(request).execute();
+        } catch (IOException e) {
+            log.error("Error while making API call to RabbitMQ");
+        }
+        return null;
```
Comment (r0goyal) [inline src/test/java/io/appform/dropwizard/actors/connectivity/actor/NamespacedQueuesTest.java:262]: When would code reach this statement ?

Comment (SwapnilBGaikwad) [reply src/test/java/io/appform/dropwizard/actors/connectivity/actor/NamespacedQueuesTest.java:262]: When rmq API call fails.

Comment (SwapnilBGaikwad) [reply src/test/java/io/appform/dropwizard/actors/connectivity/actor/NamespacedQueuesTest.java:262]: For cases where API call to RMQ fails returning null and added assertion of not null response in the test case.

File: src/test/java/io/appform/dropwizard/actors/connectivity/actor/RMQTest.java (hunk 3 above)
Comment (SwapnilBGaikwad) [reply src/test/java/io/appform/dropwizard/actors/connectivity/actor/RMQTest.java]: Updated test in `NamespacedQueuesTest` and removed RMQ Test.

**Comment by SwapnilBGaikwad** [comment]
@r0goyal Have done changes in test cases as per the comment.
If looks okay please review and merge.
cc @santanusinha

**Pull request merged**